crates/shared/src/current_block.rs

```diff
@@ -114,12 +113,53 @@ where
             .context("failed to get current block number")?
             .as_u64())
     }
+
+    /// get latest `length` blocks
+    /// if successful, function guarantees `length` blocks in Result (does not return partial results)
+    /// `last_block` block is at the end of the resulting vector
+    async fn preceding_blocks(&self, last_block: u64, length: u64) -> Result<Vec<BlockNumberHash>> {
+        ensure!(
+            length > 0 && last_block + 1 >= length,
+            "invalid input, last block {}, length {}",
+            last_block,
+            length
+        );
+
+        let include_txs = helpers::serialize(&false);
+        let mut batch_request = Vec::with_capacity(length as usize);
+        for i in (0..length).rev() {
+            let num =
+                helpers::serialize(&BlockNumber::Number((last_block.saturating_sub(i)).into()));
+            let request = self
+                .transport()
+                .prepare("eth_getBlockByNumber", vec![num, include_txs.clone()]);
+            batch_request.push(request);
+        }
+
+        // send_batch guarantees the size and order of the responses to match the requests
+        self.transport()
+            .send_batch(batch_request.iter().cloned())
+            .await?
+            .into_iter()
+            .map(|response| match response {
+                Ok(response) => serde_json::from_value::<web3::types::Block<H256>>(response)
+                    .context("unexpected response format")
+                    .and_then(|response| {
+                        Ok((
+                            response.number.context("missing block number")?.as_u64(),
+                            response.hash.context("missing hash")?,
+                        ))
+                    }),
+                Err(err) => Err(anyhow!("web3 error: {}", err)),
+            })
+            .collect()
```

Comment:
I'd like to have an `ensure` that the collected vector has the expected length to protect against nodes not following the spec.

Reply:
My understanding is that:
1. batch_response has the same number of elements as batch_request, guaranteed by https://github.com/cowprotocol/services/blob/69e80224cfb8b13a9e1cff3c896ec840ceb9cc7c/crates/shared/src/transport/http.rs#L185 if we are using http transport (which we do)
2. Final result of the function converts `Vec<Result<T>>` to `Result<Vec<T>>` returning global `Ok` if all inner results are `Ok`, otherwise returning `Err`.

I do agree one additional ensure is on a safe side, it's just I thought it is redundant based on these assumptions.

Reply:
But since we don't know which transport we are using in this function, I guess adding `ensure` here is preferable.

Reply:
That's a fair point about the batch request. I think either way is fine then. I'm ok with the extra ensure but it's not required.